### ISSUE TITLE
Algorithm wish list

### ISSUE THREAD

**Issue by perimosocordiae** [all-umass/metric-learn#13]
Methods that we haven't implemented yet, but would like to:

 - [ ] [Probablistic Global Distance Metric Learning (PGDM)](http://ai.stanford.edu/~ang/papers/nips02-metric.pdf)
  * [reference implementation](http://www.cs.cmu.edu/~epxing/papers/Old_papers/code_Metric_online.tar.gz)
 - [ ] [Discriminative Component Analysis (DCA)](http://ieeexplore.ieee.org/iel5/10925/34374/01641007.pdf)
  * [reference implementation](http://www.cs.cmu.edu/~liuy/dca.zip)
 - [ ] [Localized Distance Metric Learning (LDM)](http://www.cse.msu.edu/~yangliu1/aaai2006-distance-v7.pdf)
  * [reference implementation](http://www.cs.cmu.edu/~liuy/ldm_scripts_2.zip)
 - [ ] [DistBoost](http://www.cs.huji.ac.il/~tomboy/papers/RCA_ICML03.pdf) and [KernelBoost](http://www.cs.huji.ac.il/~tomboy/papers/KernelBoost_ICML06.pdf)
  * [reference implementation](http://www.cs.huji.ac.il/~tomboy/code/DistBoost.zip)
 - [ ] [Active Distance Metric Learning](http://www.cse.msu.edu/~yangliu1/uai2007_bayesian.pdf)
  * [reference implementation](http://www.cs.cmu.edu/~liuy/script_bayesian_dist.zip)
 - [ ] [Metric Learning for Kernel Regression (MLKR)](http://www.cs.cornell.edu/~kilian/papers/weinberger07a.pdf)
  * [reference implementation](http://www.cse.wustl.edu/~kilian/code/files/MLKR0.9.zip)

**Comment by bhargavvader** [comment]
Hello, any tips or further documentation on how to go about doing this? And which one of these would be highest priority?

**Comment by perimosocordiae** [comment]
I don't personally have a priority, but if you want to rank them you could look at their citation counts on Google Scholar. That should give you an idea of what's more popular in the literature. I would start with one that seems easy to implement, though.

Each method should have its own Python file in the `metric_learn` directory, which defines a class that inherits from `BaseMetricLearner`. This subclass should implement a `.fit()` method, as well as one of the `.metric()` or `.transformer()` methods. As a really simple example, you can start with the [`Covariance` class](https://github.com/all-umass/metric-learn/blob/master/metric_learn/covariance.py).

**Comment by perimosocordiae** [comment]
Once you have something that works, add a class to [test/metric_learn_test.py](https://github.com/all-umass/metric-learn/blob/master/test/metric_learn_test.py) which exercises the new algorithm. You should also verify that your implementation matches the reference code, if possible.

**Comment by bhargavvader** [comment]
Thanks for all the details! The paper by Xing, et al is I think the most popular (and one of the first in Metric Learning) algorithm/paper among the list, I'll give it a shot.

**Comment by dsquareindia** [comment]
Hey @perimosocordiae can I take up the MLKR implementation? I was thinking of using the `transformer()` method to use MLKR for dimensionality reduction and hence return `X_new` as the output. Also, I was thinking of having the `A` matrix stored as an attribute after the `fit()` operation thus enabling the user to access it independently also if needed. Does this seem fine to you?

**Comment by perimosocordiae** [comment]
@dsquareindia Sure, MKLR is a good choice. The `transformer()` method should return the matrix `A` (as defined the paper), and the `transform()` method will take care of the dimension reduction without needing to be overridden. In this way, the base class [`metric()` method](https://github.com/all-umass/metric-learn/blob/45c0d16aece0290ebaed3a2ef03d934526a3240f/metric_learn/base_metric.py#L8) will also work as intended (corresponding to `M` in the paper).

**Comment by aschuman** [comment]
KISSME metric learning would be a nice addition (no optimization).
You can find a reference Matlab implementation and the paper here: https://lrs.icg.tugraz.at/research/kissme/
There is also a python implementation (which I have not verified) here: https://github.com/Cysu/dgd_person_reid/blob/master/eval/metric_learning.py

**Comment by perimosocordiae** [comment]
@aschuman thanks, I've added it to the list.

**Comment by anirudt** [comment]
[LDML](http://scholar.google.co.in/scholar_url?url=https://hal.inria.fr/docs/00/43/92/90/PDF/GVS09.pdf&hl=en&sa=X&scisig=AAGBfm0pYXCDbuBhl6D6L5poN5GZXo-_4w&nossl=1&oi=scholarr&ved=0ahUKEwjikdLIu73PAhWHvo8KHZGOBQEQgAMIHCgAMAA)

**Comment by RishabGoel** [comment]
hi @perimosocordiae 
I would like to take up Probabilistic Global Distance Metric Learning.
thanks :)

**Comment by perimosocordiae** [comment]
@anirudt Thanks, added.

@RishabGoel Go for it! I look forward to your pull request.

**Comment by bhargavvader** [comment]
Hello, just wanted to know where the short form PGDM came up - I can't seem to find it in the paper linked (which is the paper just called "Distance metric learning, with application to clustering with side-information"), by Xing et al.

**Comment by RishabGoel** [comment]
@bhargavvader  Probably the wrong link. The metric is a non probabilistic one in the link. But find the PRobabilistic one in the following link:
http://www.cs.cmu.edu/~liuy/frame_survey_v2.pdf
I am bit busy right now so I am not able to do much, but will submit a PR soon.
Hope this helps!
Regards

**Comment by perimosocordiae** [comment]
@bhargavvader, @RishabGoel: yep, that's correct. I pulled the citation from the survey paper without looking too closely at it. I'll fix the link now.

**Comment by Callidior** [comment]
The method by Xing et al. (2003) is also sometimes referred to as "MMC", e.g., in the [LMNN paper](http://www.jmlr.org/papers/volume10/weinberger09a/weinberger09a.pdf) by Weinberger et al. I guess, "MM" stands for "Mahalanobis Metric Learning" and "C" for "Clustering".

**Comment by souravsingh** [comment]
@perimosocordiae I would like to take up ``DistBoost`` and ``KernelBoost``

**Comment by perimosocordiae** [comment]
@souravsingh Great! Feel free to open a PR whenever you'd like to get feedback. I just noticed the paper links for those methods are broken, so if you find working links to the papers let me know.

**Comment by souravsingh** [comment]
@perimosocordiae I think these are the links to the papers-

DistBoost- http://www.cs.huji.ac.il/~daphna/papers/distboost-icml.pdf
KernelBoost- http://leibniz.cs.huji.ac.il/tr/919.pdf

**Comment by perimosocordiae** [comment]
@souravsingh thanks, I've updated the list with your links.

**Comment by bellet** [comment]
Another suggestion that we could: the triplet-based method proposed in the paper below (heavily cited, very useful for information retrieval applications):
https://www.cs.cornell.edu/people/tj/publications/schultz_joachims_03a.pdf

Given that we currently do not have any triplet-based weakly supervised approach I think we could make this a priority. The implementation should not be very difficult

**Comment by hansen7** [comment]
How about add the extension to the existed [LSML](http://metric-learn.github.io/metric-learn/generated/metric_learn.LSML.html#metric_learn.LSML) algorithm, in the original [paper](http://web.cs.ucla.edu/~weiwang/paper/ICDM12.pdf) and [code](https://gist.github.com/kcarnold/5439917), there is the extension of the vanilla algorithm called **spLSML**, which seems to be is just to add a L1 norm of the learned Mahalanobis matrix in the loss function, thus with a novel gradient update method, it should be quick to implement.
